plugin/trino-openlineage/src/test/java/io/trino/plugin/openlineage/TestMarquezIntegration.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.openlineage;
+
+import com.google.common.collect.ImmutableMap;
+import io.airlift.units.Duration;
+import io.trino.testing.AbstractTestQueryFramework;
+import io.trino.testing.QueryRunner;
+import org.intellij.lang.annotations.Language;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.util.Locale;
+
+import static io.trino.plugin.openlineage.OpenLineageListenerQueryRunner.createOpenLineageRunner;
+import static io.trino.testing.assertions.Assert.assertEventually;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestMarquezIntegration
+        extends AbstractTestQueryFramework
+{
+    private static MarquezServer server;
+    private static String marquezURI;
+    private static final String trinoURI = "http://trino-integration-test:1337";
+    private static final HttpClient client = HttpClient.newHttpClient();
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        server = closeAfterClass(new MarquezServer());
+        marquezURI = server.getMarquezUri().toString();
+
+        return createOpenLineageRunner(ImmutableMap.<String, String>builder()
+                        .put("openlineage-event-listener.transport.type", "HTTP")
+                        .put("openlineage-event-listener.transport.url", server.getMarquezUri().toString())
+                        .put("openlineage-event-listener.trino.uri", trinoURI)
+                        .buildOrThrow());
+    }
+
+    @Test
+    void testCreateTableAsSelectFrom()
```

Comment:
We need some additional tests like creating table from view/stale materialized view/fresh MV and insert for similar operation.

Reply:
as this is an integration test I don't fully understand why we need explicit test for different data objects (tables,views,materialized views) - marquez doesn't recognize difference between trino data entities so there is no reason it would fail for particular one

Reply:
We need to ensure how the events are generated and linked if they are from different objects - like in case of `CREATE` do we map them to view or underlying table referenced table.

Reply:
ah you mean if we run CREATE VIEW VIEW AS SELECT * FROM TABLE TABLE_Y and then CREATE TABLE TABLE_X AS SELECT * FROM VIEW_A whether lineage info in inputs shows VIEW_A or TABLE_Y?

Reply:
if so - wouldn't it more fall into testing whatever fills out `QueryIOMetadata` object?

Reply:
Yeah - But we also need to ensure that - we use them in a correct way.

Reply:
We need to apply a filter on `directlyReferenced` to refer only actual tables instead of objects being referenced by views or MVs.

Reply:
but from my investigation:
- queryMetadata.getTables() won't show actual table if the view was used (and there is no reference chain present)
- queryIOMetadata.getInputs() will show tables referenced by views (and not list views at all)

so getMetadata().getTables() is not useful here because:
- if a view X of table A was used to create table B, then table A **won't be present in queryEventCompleted().getMetadata().getTables() at all** (the view will be though), but it will be present in queryCompletedEvent().getIoMetadata().getInputs() which is what we are already doing.

so how would I use directlyReferenced in such scenario?

Reply:
> if a view X of table A was used to create table B, then table A won't be present in queryEventCompleted().getMetadata().getTables() at all (the view will be though), but it will be present in queryCompletedEvent().getIoMetadata().getInputs() which is what we are already doing.

In case of populating the lineage - we need to map only the objects referred by a query right ? For a query like 
`CREATE TABLE B AS SELECT * FROM X`

 When we are looking at lineage events or the i/p required for `B` it should point only to `X` not `A` which is referred by `X` 

`queryMetadata.getTables()` should be even track the table `A` but the only difference is that `directlyReferenced` should be false in that case

Reply:
OpenLineage does not want to know about views, only tables. So if in your example B is a table, X is a view and A is a table (upon which view X is created), the lineage we want is:
- inputs: A
- outputs: B

this is by design

Reply:
In case `OpenLineage` we are tracking tables and columns and how are they linked, but in case of SQL views/MVs are also table like object right - so shouldn't we consider a view/MV as a part of lineage. For instance in case of Apache Atlas (a tool which tracks the lineage) we would treat View/MVs as a table so any PII or tags could be set on them as well.

Reply:
@mobuchowski could you chip in here? is there a place in OL where we reference views instead of tables? or - if a table is created based on a view should be store table as input and this view as a symlink?

Reply:
I'm not a OpenLineage expert but Table/View/MV are similar relation objects which we use in a query - though it could be resolved differently in the runtime - the query string which is being executed doesn't capture those abstraction - like `SELECT * FROM view_1` - Doesn't mention any details any information about `view_1` - it is being abstracted and the lineage should point to the abstraction instead of underlying references.

In case of Trino `SHOW TABLES` would treat everything as table object even if it is a view or a MVs.

Reply:
yes but imagine a scenario in which you do

- `CREATE VIEW VIEW_X AS SELECT * FROM TABLE_A`
- `CREATE TABLE TABLE_B AS SELECT * FROM VIEW_X`

in this scenario, if we have
- inputs: VIEW_X
- outputs: TABLE_B

then when someone removes VIEW_X from Trino, the lineage is gone as well - which is wrong. Even if VIEW_X is gone we should have lineage information present as actual flow of data was from TABLE_A.

Reply:
But this problem is irrespective of View/Table/MV right ? 

- `CREATE TABLE TABLE_X AS SELECT * FROM TABLE_A`
- `CREATE TABLE TABLE_B AS SELECT * FROM TABLE_X`

If someone drop `TABLE_X` then the lineage is cut off right `TABLE_B` would not be able to point to `TABLE_A` ? Ideally when we delete an entity it should be marked as inactive and the lineage should be maintained for historic purpose.

Reply:
In ideal situation you'd get lineage from all parts: 
```
CREATE VIEW VIEW_X AS SELECT * FROM TABLE_A
```
would produce 
 - inputs: TABLE_A
 - outputs: VIEW_X

and as you said, 
```
CREATE TABLE TABLE_B AS SELECT * FROM VIEW_X
```
would produce 
 - inputs: VIEW_X
 - outputs: TABLE_B

> then when someone removes VIEW_X from Trino, the lineage would be gone as well - which is wrong. Even if VIEW_X is gone we should have lineage information present as actual flow of data was from TABLE_A.

I think this is implementation detail of data catalog (or other tool) that consumes OL events, I don't see why removal should result from immediate deletion of lineage?

Reply:
this is not the case for spark however, the views are translated into tables and tables are referenced in inputs. am I right?

Reply:
ok, I refactored the code so it:
- does not use queryCompletedEvent.getIOMetadata() at all
- uses queryCompletedEvent.getMetadata().getTables()
- inputs section is populated by entities found in getTables which have `isReferencedDirectly` set to `true`
- tables that are marked as not `isReferencedDirectly` go into symlinks.identifiers for corresponding table through referenceChain mapping

in other words:
1. on CREATE TABLE TABLE_B AS SELECT * FROM TABLE_A we should have:
- inputs: TABLE_A
- outputs: TABLE_B

2. on CREATE TABLE TABLE_C AS SELECT * FROM VIEW_A we should have:
- inputs: VIEW_A (plus symlinks.identifiers=TABLE_C)
- outputs: TABLE_B

let me know if that's what you both had in mind.

Reply:
```
  ~ curl localhost:32813/api/v1/namespaces/trino%3A%2F%2Ftrino-integration-test%3A1337/datasets/marquez.default.test_view | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1352  100  1352    0     0  10770      0 --:--:-- --:--:-- --:--:-- 10816
{
  "id": {
    "namespace": "trino://trino-integration-test:1337",
    "name": "marquez.default.test_view"
  },
  "type": "DB_TABLE",
  "name": "marquez.default.test_view",
  "physicalName": "marquez.default.test_view",
  "createdAt": "2024-04-24T13:47:33.895Z",
  "updatedAt": "2024-04-24T13:47:33.895Z",
  "namespace": "trino://trino-integration-test:1337",
  "sourceName": "default",
  "fields": [
    {
      "name": "nationkey",
      "type": null,
      "tags": [],
      "description": null
    },
    {
      "name": "regionkey",
      "type": null,
      "tags": [],
      "description": null
    },
    {
      "name": "name",
      "type": null,
      "tags": [],
      "description": null
    },
    {
      "name": "comment",
      "type": null,
      "tags": [],
      "description": null
    }
  ],
  "tags": [],
  "lastModifiedAt": null,
  "lastLifecycleState": "",
  "description": null,
  "currentVersion": "f9e2368e-6d1e-4b33-b6a2-2a61606f38ae",
  "columnLineage": null,
  "facets": {
    "schema": {
      "fields": [
        {
          "name": "nationkey"
        },
        {
          "name": "regionkey"
        },
        {
          "name": "name"
        },
        {
          "name": "comment"
        }
      ],
      "_producer": "https://github.com/trinodb/trino/plugin/trino-openlineage",
      "_schemaURL": "https://openlineage.io/spec/facets/1-1-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet"
    },
    "symlinks": {
      "_producer": "https://github.com/trinodb/trino/plugin/trino-openlineage",
      "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SymlinksDatasetFacet.json#/$defs/SymlinksDatasetFacet",
      "identifiers": [
        {
          "name": "tpch.tiny.nation",
          "type": "TABLE",
          "namespace": "trino://trino-integration-test:1337"
        }
      ]
    }
  },
  "deleted": false
}
```

Reply:
Looks great!